ondemand/README.md

```diff
@@ -1,4 +1,389 @@
+developing the application and starting new jobs.
+
+#### Save your spot
+
+Now it's probably a good idea to save the modifications. They're small, but it'll still help if you
+ever get into trouble and need to revert. A simplified version of the `form.yml` is in the very next
+section, and you may want to use and save _it_ instead so that any `git diff` you do will be much
+smaller and easier to read.
+
+You can use the
+[shell app to login to this directory](https://localhost:3443/pun/sys/shell/ssh/default/home/hpcadmin/ondemand/dev/jupyter/)
+
+In this shell you'll save in git with these commands:
+
+```shell
+git add .
+git commit -m 'initial commit that correctly submits to the hpc cluster'
+```
+
+### A closer look at the form.yml
+
+The items in the form.yml directly create what's shown to the users in the form they interact with.
+Let's take a closer look at the `form.yml` that created the form you just submitted to get an
+understanding of how they relate to what's shown in the UI.
+
+This is the `form.yml` you at this point without all the comments.
+```yaml
+cluster: "hpc"
+attributes:
+  modules: "python"
+  extra_jupyter_args: ""
+form:
+  - modules
+  - extra_jupyter_args
+  - bc_account
+  - bc_queue
+  - bc_num_hours
+  - bc_num_slots
+  - bc_email_on_started
+```
+
+All fields pre-pended with `bc_` are special fields OnDemand provides for convenience. They are commonly
+used fields that create corresponding script attribute.  We'll talk more about script attributes later.
+
+* `modules` Specifies the modules loaded. Since it's hard coded to "python" (in the attributes) 
+    we didn't see it in the form.
+* `extra_jupyter_args` Specifies the extra jupyter arguments but since it's hard coded to "" we didn't
+    didn't see it in the form.
+* `bc_account` Creates the "Account" text field and submits the job with the given account.
+* `bc_queue` Creates the "Partition" text field and submits the job to the given partition.
+* `bc_num_hours` Creates the "Number of hours" integer field and submits the job with the given
+    walltime.
+* `bc_num_slots` Creates the "Number of nodes" integer field and submits the job with the requested
+    cores.
+* `bc_email_on_started` Creates the "I would like to receive an email when the session starts" checkbox
+    and submits the job with a request to email when the job starts.
+
+### Updating the Jupyter App
+
+#### Change bc_queue to a select field
+
+We have 2 partitions enabled in the SLURM containers (SLURM calls queues partitions, so we'll be switching
+back and forth between the two terms in this tutorial). We've started with a field `bc_queue` that
+is a text field, but it's likely much easier for users to simply choose the partition out of a
+select dropdown menu instead.
+
+So let's remove the `bc_queue` field in the form and replace it with a new field we'll call `custom_queue`.
```

Comment:
`partition` instead of `custom_queue` - since we are using Slurm, why not?

Reply:
I can change it and quickly show that the Partition displays as the default label. Then I can show if you add label and change it to "Choose Partition" how it changes.

Reply:
Yea, I tried partition initially, but it's being created as an enumerator.  I can't figure out why, I guess some rails thing.  

 If i just replace `- bc_queue` with `- partition` this is what shows up. Then SLURM barfs in the actual submit because that's the string that's submitted even when I change the attributes to a select widget and give the right options

I can't figure out why, especially just as `custom_queue` directly works.
![image](https://user-images.githubusercontent.com/4874123/87821935-a4679200-c83e-11ea-9b08-08e2d8b62797.png)

Reply:
Partition is a function on Enumerable, that's why. Wonder why nobody has mentioned this yet?

https://ruby-doc.org/core-2.5.1/Enumerable.html#method-i-partition